Fix #7921 and added test cases

ayeminag opened the pull request:
### NOTES
In order to run `internal/e2e-realtime-api/voiceSpeechCollect.test.ts` before merging the backend changed
#### Backend Todos
- setup devbox with FS changes
- setup steering to devbox
- create a sip domain app on staging and handle the call on domain app with a relay context `relaye2e`
- add routing to devbox in kamailio for the domain app you set up in previous step

#### SDK Todos
- in `internal/e2e-realtime-api/src/voiceSpeechCollect.test.ts`, replace `dev-min.dapp.swire.io` sip domain with the sip domain you created in above step.
- copy env file into `internal/e2e-realtime-api/.env.test` and setup 
- run `npm dev:rt-only`
### References
- [#7921](https://github.com/signalwire/cloud-product/issues/7921)

### Description
- Added `state` param to `CallingCallCollectEventParams`
- Made sure `voiceCallCollectWorker` doesn't clean up `CallCollect` instance and emit ended/failed event if the `state` is `"collecting"`
- Resolve `CallCollect.ended()` promise only when state is NOT `"collecting"` AND `final` is either `undefined`/`true` AND `result.type` is one of `ENDED_STATES`.
- Added more test cases for `Call.collect()` in `@sw-internal/e2e-realtime-api`

giavac commented on internal/e2e-realtime-api/src/voiceSpeechCollect.test.ts:
```diff
@@ -0,0 +1,156 @@
+
+      try {
+        const resultAnswer = await call.answer()
+        tap.ok(resultAnswer.id, 'Inboud call answered')
```
A minor typo in log

ayeminag replied:
fixed

iAmmar7 commented on internal/e2e-realtime-api/src/voiceSpeechCollect.test.ts:
```diff
@@ -0,0 +1,199 @@
+            //   name: 'from',
+            //   domain: domainApp.domain,
+            // }),
+            to: 'sip:to@dev-min.dapp.swire.io',
```
I think you forgot to change it back to the dynamic sip domain address (`makeSipDomainAppAddress`)?

ayeminag replied:
changed it back

iAmmar7 commented on internal/e2e-realtime-api/src/voiceSpeechCollect.test.ts:
```diff
@@ -0,0 +1,199 @@
+
+          // const sendDigitResult = await call.sendDigits('1w2w3w#')
+          
+          // tap.equal(
```
Can we please remove all these unnecessary comments as well?

ayeminag replied:
Sure test is not final yet like I mentioned on slack.

iAmmar7 commented on packages/realtime-api/src/voice/CallCollect.ts:
```diff
@@ -155,8 +163,8 @@ export class CallCollectAPI
     if (
-      ENDED_STATES.includes(this.result?.type as CallingCallCollectEndState)
-    ) {
+      this.state != 'collecting' && (this.final === undefined || this.final === true)  &&
```
Considering we have collect states now: collecting | error | finished.

I think we don't need all these conditions. Let's discuss, perhaps we could just simply rely on the state. Something like this:

```javascript
const ENDED_STATES = ['error', 'finished']
if (ENDED_STATES.includes(this.state)) {
  return Promise.resolve(this)
}
```

ayeminag replied:
Since, `state` is only available in `continuous: true` case AND we also need to consider `partialResults: true`, I think we need it? @iAmmar7

iAmmar7 replied:
DISCUSSED: `this.final !== false`

iAmmar7 commented on packages/realtime-api/src/voice/workers/voiceCallCollectWorker.ts:
```diff
@@ -50,22 +50,27 @@ export const voiceCallCollectWorker = function* (
         case 'no_match':
         case 'error': {
-          callInstance.emit(`${eventPrefix}.failed`, actionInstance)
+          if (payload.state !== 'collecting') {
```
TBD

After the inclusion of the state, we might be able to reshape this whole worker and make it consistent with other workers.

ayeminag replied:
@iAmmar7 is this still true? state is only there for `calling.call.collect` events.

iAmmar7 replied:
DISCUSSED: State is only available with `continuous: true`. So, it is fine

iAmmar7 commented on packages/core/src/types/voiceCall.ts:
```diff
@@ -1006,12 +1006,14 @@ export type CallingCallCollectEndState = Exclude<
   control_id: string
   result: CallingCallCollectResult
   final?: boolean
+  state?: CallingCallCollectState
```
State is not optional, is it?

ayeminag replied:
State is only present when the collect was started with `continuous: true`

ayeminag commented:
@giavac I added the steps we need to take to run test before backend changes are on staging.

giavac commented on internal/e2e-realtime-api/src/voiceSpeechCollect.test.ts:
```diff
@@ -0,0 +1,201 @@
+            //   name: 'from',
+            //   domain: domainApp.domain,
+            // }),
+            to: `sip:to-${uuid()}@dev-min.dapp.swire.io`,
+            from: `sip:from-${uuid()}@dev-min.dapp.swire.io`,
```
Marking as part of the feedback that these changes must be removed before merging.

ayeminag replied:
removed @giavac

giavac reviewed (changes requested):
Bookmark to ensure the generic SIP Domains are restored before merging.

giavac commented:
(I'll approve once the other open conversations will be resolved)

ayeminag commented:
I updated the tests to make sure to fail on prod, but it seems like we deployed the collect changes to production already. 
But if the state is missing in continuous: true case, SDK will just resolve the `collect.ended()` and cleanup `CallCollect` instance on first collect event with `final: true` (in case of `partialResults: true`) OR on first speech type collect event (in case `partialResults: false | undefined`). And when new events come in after that sdk will throws  `Error('Missing the Instance')`, sdk is already catching that error and logging it to stderr. So, I added tests to make sure Missing the Instance error wasn't logged to stderr, if the event is missing state in `continuous: true` case tests will fail. 

I updated the audio file I am playing in the test to include long enough pause, so FS will deliver more than one collect events (at least one event with final: true in partialResults: true, continuous: true case), and adjusted silenceTimeout and speechTimeout to accommodate it. 

Since, changes are on prod, to test the failure. I disabled the `state !== 'collecting'` checks in sdk first (because not checking for `state` on sdk side is behaviorally same as `state` missing from FS side events), and ran the test, tests failed because it's getting `Missing the Instance` error just like it should. See the logs [here](https://github.com/signalwire/signalwire-js/files/13316513/e2e-test-logs.txt).

I also notice speech detect can return inconstant data for example `1 2 3 4 5` instead of  `one two three four five` I updated assert for all possible cases, so they will pass as long as it's one of the possible expected texts.

/cc @iAmmar7

iAmmar7 reviewed (approved):
LGTM. Thanks @ayeminag 🚀
